File: demo-ios/demo-ios/Message.swift
//
//  Message.swift
//  demo-ios
//
//  Created by xuduo on 12/11/14.
//  Copyright (c) 2014 yy. All rights reserved.
//

import Foundation

class Message {
    
    var command = ""
    var headers = [String: String]()
    var payload = ""
    
    func destination() -> String {
        return headers["destination"]!
    }
    
    class func toMessageString(command:String,headers:[String:String],payload:String) -> String{
        var result = command + "\n"
        
        if(headers.count > 0){
            for (key,value) in headers {
                result += key + ":" + value + "\n"
            }
        }
        
        result += "\n"
        
        result += payload
        
        result += "\0"
        
        println("toMessageString\n\(result)")
        
        return result
        
    }
    
    class func parse(result:String) -> Message {
        println("Message.parse")
        var message = Message()
        var reader = StringReader(string: result)
        message.command = reader.readline()
        
        var header = ""
        while true {
          header = reader.readline()
            println("header readLine + \n\(header)")
          if header == "" || header == "\0" {
                break
           }
          var array = header.componentsSeparatedByString(":")
          message.headers[array[0]] = array[1]
        }
        
        
        
        while true {
            var char = reader.read()
            if(char == "\0"){
                break
            }
            message.payload += char
        }
        
        return message
    }
    

    
}

File: stomp-android/src/main/java/com/yy/androidlib/websocket/Message.java
package com.yy.androidlib.websocket;

import android.util.Log;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.StringReader;
import java.util.HashMap;
import java.util.Map;

public class Message {

    private String command;
    private Map<String, String> headers = new HashMap<String, String>();
    private String body;

    public static Message parse(String data) {
        data = data.replace("\\n", "\n");
        data = data.replace("\\u0000", "\000");
        data = data.replace("\\", "");
        Message message = new Message();
        try {

            BufferedReader reader = new BufferedReader(new StringReader(data));
            message.command = reader.readLine();
            String header;
            while ((header = reader.readLine()).length() > 0) {
                int ind = header.indexOf(':');
                String k = header.substring(0, ind);
                String v = header.substring(ind + 1, header.length());
                message.headers.put(k.trim(), v.trim());
            }
            StringBuffer body = new StringBuffer();
            int b;
            while ((b = reader.read()) != 0) {
                body.append((char) b);
            }
            message.body = body.toString();

        } catch (IOException e) {
            Log.e("STOMP", "parse message fail", e);
        }
        return message;
    }

    public String getCommand() {
        return command;
    }

    public void setCommand(String command) {
        this.command = command;
    }

    public Map<String, String> getHeaders() {
        return headers;
    }

    public void setHeaders(Map<String, String> headers) {
        this.headers = headers;
    }

    public String getBody() {
        return body;
    }

    public void setBody(String body) {
        this.body = body;
    }

    public String getDestination() {
        return getHeaders().get("destination");
    }

    public static String toRawString(Command c, String[] h, String b) {
        StringBuffer message = new StringBuffer(c.toString());
        message.append("\n");

        if (h != null) {
            for (int i = 0; i + 1 < h.length; i = i + 2) {
                message.append(h[i]);
                message.append(":");
                message.append(h[i + 1]);
                message.append("\n");
            }
        }

        message.append("\n");


        if (b != null) message.append(b);

        message.append("\000");

        String data = "[" + JSONObject.quote(message.toString()) + "]";

        return data;
    }

}
